### PR TITLE
Bte 130

### PR DESCRIPTION
**Pull request by BruceChar**
## Context

* Describe the changes made and why they were made.
Context:
 parallel service relys on subquery once the subquery network is down that parallel service api will be unavailable. 
Resolution:
 Add cache to prevent from api unavailable, it shouldn't crash if parallel service cannot fetch the latest data from third service.
* Provide links to any issues or designs for this change.

## Were any new external packages added?
* List external packages added and why they were added
aws-sdk ^2.1094.0, for cloudwarch alert

### CODE REVIEW

**Comment by BruceChar** [comment]
depart the pr into two pieces, 1. the redis util wrap restruct and add the cache file; 2. cache the subquery request

File: packages/app/package.json [hunk 1]
```diff
@@ -31,7 +31,8 @@
     "reflect-metadata": "^0.1.13",
     "typeorm": "^0.2.44",
     "winston": "^3.6.0",
-    "winston-daily-rotate-file": "^4.6.1"
+    "winston-daily-rotate-file": "^4.6.1",
+    "aws-sdk": "^2.1094.0"
```
Comment (crystalynliu) [inline packages/app/package.json:35]: Remove the aws sdk dependencies

File: packages/app/src/service/redis.ts [hunk 2]
```diff
@@ -1,13 +1,19 @@
...
 import { EventQuery } from '../subql/event'
-import { AssetMeta, MarketMeta } from '../types'
+import { AssetMeta, MarketMeta, NetworkT } from '../types'
+import KEYS from '../redisKeys'
 
 import { QueryApi } from './query'
 
 const log = getAppLogger('redis')
 
+enum DBT {
```
Comment (crystalynliu) [inline packages/app/src/service/redis.ts:12]: what's this enum for? Why do you not merge the `DBT { Graphql = 3 }` in one place?

File: packages/app/src/types/index.ts [hunk 3]
```diff
@@ -1,2 +1,3 @@
+export type NetworkT = 'PARA' | 'HEIKO' | 'heiko' | 'para'
```
Comment (crystalynliu) [inline packages/app/src/types/index.ts:1]: why to support four network here?

File: packages/app/src/utils/cloudwatch.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import { CloudWatch } from 'aws-sdk'
+import type { MetricData } from 'aws-sdk/clients/cloudwatch'
+import { getAppLogger } from '@parallel/lib'
+
+const log = getAppLogger('cloudwatch')
+const CLOUD_WATCH_NS = 'parallel-service'
+
+const cloudwatchClient = new CloudWatch()
+
+export enum CloudWatchError {
```
Comment (crystalynliu) [inline packages/app/src/utils/cloudwatch.ts:10]: Maybe remove the cloudwatch for now, we can create another card to do it

File: packages/app/src/service/redis.ts (hunk 2 above)
Comment (BruceChar) [reply packages/app/src/service/redis.ts:12]: it should be!

File: packages/app/src/types/index.ts (hunk 3 above)
Comment (BruceChar) [reply packages/app/src/types/index.ts:1]: eh! It's confused that only two network given four, I'll modify the usecase

**Review by xsteadybcgo — CHANGES_REQUESTED** [review]

File: packages/lib/src/redis.ts [hunk 5]
```diff
@@ -1,17 +1,8 @@
 import IORedis, { RedisOptions } from 'ioredis'
 
-export enum DBT {
-  User = 0,
-  Cache = 1
-}
-
-export type NetworkType = 'para' | 'heiko'
+type DBT = number
```
[the remark this answers is not in the document]
Comment (BruceChar) [reply packages/lib/src/redis.ts:3]: it's in lib package just a alias

File: packages/app/src/utils/config.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  static getRedisConfig(network: NetworkT = 'heiko'): RedisConfig {
+    return {
+      host: getStringEnv(`${network.toUpperCase()}_REDIS_HOST`),
+      port: getNumEnv('REDIS_PORT'),
+      username: getStringEnv('REDIS_USER'),
+      password: getStringEnv('REDIS_PASSWORD')
+    }
+  }
+
+  static getEndpointsConfig(network: NetworkT = 'heiko'): EndpointConfig {
```
[the remark this answers is not in the document]
Comment (BruceChar) [reply packages/app/src/utils/config.ts:46]: yes! for the database branch

File: packages/lib/src/redis.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (xsteadybcgo) [reply packages/lib/src/redis.ts:3]: I think it is better to define `DBT` in the `@parallel/lib/src/redis.ts`,  and import this type from the '@parallel/lib'

[the remark this answers is not in the document]
Comment (BruceChar) [reply packages/lib/src/redis.ts:3]: the db is related to business, it cannot identify which db to use previously until we have a new requirement

**Review by xsteadybcgo — APPROVED** [review]

**Pull request merged**